sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -1257,6 +1285,8 @@ inline std::string GetConnectionKey(std::string const& host, CurlTransportOption
   std::string key(host);
   key.append(!options.CAInfo.empty() ? options.CAInfo : "0");
   key.append(options.Proxy ? (options.Proxy->empty() ? "NoProxy" : options.Proxy.Value()) : "0");
+  key.append(options.ProxyUsername.empty() ? "0" : options.ProxyUsername);
+  key.append(options.ProxyPassword.empty() ? "0" : options.ProxyPassword);
```

Comment:
consider to change this at some point to another approach, as the key is getting too big already.
Probably getting a hash from everything instead of the long string.

Also, you need to also add the `EnableCurlTracing` option.  Otherwise, you might get a weird behaviour from having one client with log and another without log.  It is a corner case, but it might be annoying to create a new client and get a verbose client due to a previous step.

And last thing..  saving the user and password here might bring some security concerns, as it is becoming part of the app state. (maybe try hash() them)

Reply:
Good point, let me think about this.

I'm not worried about the logging state or password case stuff (since we already have the username and password as a member in the CurlConnection), but the complexity argument is valid. 

It might make sense to use a std::unordered_map to hold this.

Reply:
Ok, just looked. And then I screamed quietly to myself:
```c++
   /**
     * @brief Keeps a unique key for each host and creates a connection pool for each key.
     *
     * @details This way getting a connection for a specific host can be done in O(1) instead of
     * looping a single connection list to find the first connection for the required host.
     *
     * @remark There might be multiple connections for each host.
     */
    std::map<std::string, std::list<std::unique_ptr<CurlNetworkConnection>>> ConnectionPoolIndex;
```

`std::map` stores its values in an RB tree, not a hash table. So the lookup is `O(logn)`, not `O(1)`.

Reply:
shame on me.
we should change it to `std::unordered_map`

Reply:
Yup. Also looking at changing the key from std::string to a struct.

Reply:
Ok, changing the key to a struct is ... unpleasant and it's not clear it provides significant value. Looking at a different solution now.